content/blog/2020-06-28-shtab-completion.md

```diff
@@ -0,0 +1,197 @@
+
+descriptionLong: |
+  We've made a painless tab-completion script generator for Python applications!
+  It's called `shtab` and it currently works with `argparse`, `docopt`, and
+  `argopt` to produce `bash` and `zsh` completion scripts.
+  This tool was originally created to help `dvc`, but we realised it could be made
+  more generic and valuable to the world's entire ecosystem of Python CLI
+  applications. Find out how to take advantage of it in this blog post.
+author: casper_dcl
+---
+
+> Other potential titles:
+>
+> - Zero Effort Tab Completion for Python Applications
+> - Finally: 1-Click Tab Completion for [`argparse`], [`docopt`] and [`argopt`]
+> - A [`shtab`] in the dark at completing your ~~sentences~~ commands
+
+Command line tools are powerful. Things like [`make`] have manual pages
+spanning, well,
+[pages](https://www.gnu.org/software/make/manual/make.html#Options-Summary),
+while just the list of [`git`] subcommands is longer than can fit on a standard
+`80 x 24` terminal screen.
+
+```dvc
+$ git <TAB>
+add                  filter-branch        rebase
+am                   format-patch         reflog
+annotate             fsck                 relink
+...
+describe             prco                 unassume
+--More--
+```
+
+Notice the `--More--` at the bottom? That's the joy of pagination.
+
+Notice the `<TAB>` at the top? That represents actually pressing the tab key.
+Ah, the joy of shell tab completion.
+
+Tab completion is an indispensable part of writing anything on the command-line.
+Personally, I can't imaging trying to `git co` (aliased to `git checkout`) a
+branch without `<TAB>` to do the heavy lifting.
+
+Now there's a tool called `dvc` which is like [Git for data](https://dvc.org).
+It can be viewed as a cross-platform combination of [`git`] and [`make`]
+designed for handling big data and multiple cloud storage repositories, as well
+as tracking machine learning experiments. As you can imagine, supporting that
+many buzzwords means it also has a large number of subcommands and options.
+
+_Every time a new feature is added, maintainers and contributors have to update
+tab completion scripts for multiple supported shells. At best, it's a pain, and
+at worst, error-prone. If you've worked on maintaining CLI applications, you'll
+sympathise._
+
+Surely the parser code you've written is informative enough to automate tab
+completion? Surely you shouldn't have to maintain and synchronise separate tab
+completion scripts?
+
+Good news: [`shtab`] is a new tool which magically does all of this work.
+
+Any Python CLI application using [`argparse`], [`docopt`], or [`argopt`] can
+have tab completion for free!
+
+Simply hand your parser object to `shtab` (either via the CLI or the Python
+API), and a tab completion script will be generated for your preferred shell.
+It's as easy as:
+
+- CLI: `shtab --shell=bash myprogram.main.parser`, or
+- Python API: `import shtab; print(shtab.complete(parser, shell="bash"))`.
```

Comment:
Nice! Is this part of our CI process now? Just curious how we're doing it exactly.

Reply:
there are no tests at the moment... Though actually the `snap` package will run `dvc completion -s bash` internally during the build so at least we'll know if the command exits non-zero.

Reply:
OK. I meant are we generating the auto complete script for DVC with shtab during deployment, perhaps? Or is it done manually for now

Reply:
Some distributions (snap, brew, deb) automatically enable completion. For other methods (pip, conda) users can manually run `dvc completion` following https://dvc.org/doc/install/completion

Reply:
Gotcha. BTW is there a way to install on Git Bash, you think? For Windows users. There's no sudo or apt obvi.

Also! Pleas notice #1554 @casparjespersen, just need a quick approval so I can merge. Thanks

Reply:
@jorgeorpinel wrong Casper :) and we don't merge this, we'll be publishing it next week (since we have too much stuff for this one already) on dev.to most likely. So, the best way to give some feedback is to read dev.to and see what's popular and why.

Reply:
windows users can use git-bash and get completion, yes. They can follow the same instructions as Linux (install into `/etc/bash_completion.d/`). Maybe want to add that to the docs? Also someone could contribute e.g. powershell completion to the shtab repo :)